lib/babe/babe.go

```diff
@@ -336,158 +303,105 @@ func (b *Service) initiate() {
+		epochData.threshold, epochData.randomness[:], epochData.authorities)
 
-		epochStartSlot, err := b.waitForEpochStart(epoch)
-		if err != nil {
-			logger.Errorf("failed to wait for epoch %d start: %s", epoch, err)
-			return err
-		}
-
-		// calculate current slot
-		startSlot := getCurrentSlot(b.slotDuration)
-		intoEpoch := startSlot - epochStartSlot
-
-		// if the calculated amount of slots "into the epoch" is greater than the epoch length,
-		// we've been offline for more than an epoch, and need to sync. pause BABE for now, syncer will
-		// resume it when ready
-		if b.epochLength <= intoEpoch && !b.dev {
-			logger.Debugf(
-				"pausing BABE, need to sync since we have %d slots (start slot %d) into the epoch starting at %d",
-				intoEpoch, startSlot, epochStartSlot)
-			return b.Pause()
-		}
+	epochStartSlot, err := b.epochState.GetStartSlotForEpoch(epoch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start slot for current epoch %d: %s", epoch, err)
+	}
 
-		if b.dev {
-			intoEpoch = intoEpoch % b.epochLength
-		}
+	return newEpochHandler(epoch,
+		epochStartSlot,
+		epochData,
+		b.constants,
+		b.handleSlot,
+		b.keypair,
+	)
+}
 
-		logger.Infof("current epoch %d has %d slots", epoch, intoEpoch)
+func (b *Service) runEngine() error {
+	epoch, err := b.epochState.GetCurrentEpoch()
+	if err != nil {
+		return fmt.Errorf("failed to get current epoch: %s", err)
+	}
 
-		// get start slot for current epoch
-		nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-		if err != nil {
-			logger.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			return err
-		}
+	for {
+		err = func() error {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			b.epochHandler, err = b.initiateAndGetEpochHandler(epoch)
+			if err != nil {
+				return err
+			}
 
-		nextEpochStartTime := getSlotStartTime(nextEpochStart, b.slotDuration)
-		epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
-		cleanup := func() {
-			if !epochTimer.Stop() {
-				<-epochTimer.C
+			// get start slot for current epoch
+			nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
+			if err != nil {
+				return fmt.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
 			}
-		}
 
-		slotDone := make([]<-chan time.Time, b.epochLength-intoEpoch)
-		for i := 0; i < int(b.epochLength-intoEpoch); i++ {
-			slotDone[i] = time.After(b.getSlotDuration() * time.Duration(i))
-		}
+			nextEpochStartTime := getSlotStartTime(nextEpochStart, b.constants.slotDuration)
+			epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
+			cleanup := func() {
+				if !epochTimer.Stop() {
+					<-epochTimer.C
+				}
+			}
 
-		for i := 0; i < int(b.epochLength-intoEpoch); i++ {
-			done := false
+			errCh := make(chan error)
+			go b.epochHandler.run(ctx, errCh)
 
 			select {
 			case <-b.ctx.Done():
+				cancel()
```

Comment:
as we return we don't need to call the `cancel()` already defined in the `defer` statement
```suggestion
				cancel()
```

Reply:
updated